### ISSUE TITLE
Question about does mamba support variable-length input or cu_seqlens like flash attention?

### ISSUE THREAD

**Issue by zigzagcai** [state-spaces/mamba#180]
We know that flash attention supports `cu_seqlens`, which can remove padding for variable-length input in a batch and only store regular tokens. This can be useful for optimizing the computational efficiency when packing multiple short sequences.

So, does Mamba also have this mechanism such as variable-length input or `cu_seqlens` like flash attention?

**Comment by tridao** [comment]
Yes, there should be ways to deal with variable length. It's not implemented yet however.

**Comment by zigzagcai** [comment]
Got it. Thank you Tri Dao!

**Comment by zigzagcai** [comment]
Sorry but I still have some questions:

Is it theoretical possible for Mamba to provide variable-length API like Flash-Attention([here](https://github.com/Dao-AILab/flash-attention/issues/432#issuecomment-1698610752))?

**Comment by zigzagcai** [comment]
> Yes, there should be ways to deal with variable length. It's not implemented yet however.

Sorry but I still have some confusions:

Is it theoretical possible for Mamba to provide variable-length API like Flash-Attention (https://github.com/Dao-AILab/flash-attention/issues/432#issuecomment-1698610752)?
Since in most cases, for the computing efficiency, we want to concatenating short samples to get one packed sample. We know that for Transformer-based models, we can use flash-attention API which contains cu_seqlens to process packed samples.

**Comment by zigzagcai** [comment]
From my understanding, since `conv1d` and `parallel associative scan` in the Mamba block are linear operations, the processing of pack data can theoretically be achieved through mask. 
Not sure if my understanding is correct?

**Comment by zigzagcai** [comment]
Just have another question, could Mamba be parallelized over `seq_len` like the flash-attention implementation of Transformer?

**Comment by albertfgu** [comment]
In practice, depending on your setting, you may be able to simply concatenate the sequences and pass the whole sequence in (without enforcing state resetting at sequence boundaries). I've used this in the past where it has worked fine in some settings.

**Comment by deroholic** [comment]
> In practice, depending on your setting, you may be able to simply concatenate the sequences and pass the whole sequence in (without enforcing state resetting at sequence boundaries). I've used this in the past where it has worked fine in some settings.

It is often done that way, but it does cause sample cross contamination during training and that is usually not desirable.

**Comment by albertfgu** [comment]
Yes. I'm just saying sometimes it's also fine :)

**Comment by zigzagcai** [comment]
> In practice, depending on your setting, you may be able to simply concatenate the sequences and pass the whole sequence in (without enforcing state resetting at sequence boundaries). I've used this in the past where it has worked fine in some settings.

Does that mean selective SSM can automatically reset boundaries like attention mask in Transformer models?
I see in the section 3.5.2 of Mamba paper and find below description.
![image](https://github.com/state-spaces/mamba/assets/8370601/ff40ff72-c7b7-45bd-91f5-0280c827f9a9)

**Comment by zigzagcai** [comment]
I also see one blog on together.ai, where the next steps shows that variable length training are on the future roadmap.
It would be fantastic if mamba could provides this feature like transformer in the future!
![image](https://github.com/state-spaces/mamba/assets/8370601/c8aeb963-a1f2-4d3d-bf1a-b5381872c57f)

**Comment by zigzagcai** [comment]
Has been done in https://github.com/state-spaces/mamba/pull/244
